src/middleware/cors/service.rs

```diff
@@ -13,27 +16,196 @@ pub struct CorsService<S> {
 
 impl<S> CorsService<S> {
     pub(super) fn new(inner: S, config: Arc<Config>) -> CorsService<S> {
-        CorsService {
-            inner,
-            config,
-        }
+        CorsService { inner, config }
     }
 }
 
 impl<S, RequestBody, ResponseBody> Service for CorsService<S>
-where S: Service<Request = http::Request<RequestBody>,
-                Response = http::Response<ResponseBody>>,
+where
+    S: Service<Request = http::Request<RequestBody>, Response = http::Response<ResponseBody>>,
+    ResponseBody: BufStream,
 {
-    type Request = S::Request;
-    type Response = S::Response;
+    type Request = http::Request<RequestBody>;
+    type Response = http::Response<Option<ResponseBody>>;
     type Error = S::Error;
-    type Future = S::Future;
+    type Future = CorsFuture<S::Future>;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        // what are we polling
```

Comment:
@carllerche I forgot to ask you this question. What is it that we are polling on a *`Service`*?

Reply:
Any resources, if any, that would be needed to answer the request. For middleware, it'd likely be the wrapped service. But it could also be other things, like how tower-buffer will return NotReady once the buffer limit is reached.